### DISCUSSION TITLE
Idea: using pydantic in v1.0

### DISCUSSION THREAD

**Discussion by grutts** [mobilityhouse/ocpp#182] [Ideas]
I was thinking about deserialisation of incoming requests to dataclasses in [this PR](https://github.com/mobilityhouse/ocpp/pull/172) - in particular nested dictionaries. Starting a new discussion here as it's a separate scope. 

It would be possible for us to implement such deserialisation in the library. However, pydantic models (and pydantic dataclasses) already support this, and using them as the primary data structure in v1.0 is generally a good fit.

1. constructing nested classes from nested dictionaries is already supported by pydantic
2. pydantic models validate the data types, which is handy for this library's use case
3. pydantic models can be auto-generated from json schemas, which is also handy for this library's use case

What do you think? 

Pydantic's dataclasses are very similar to normal python dataclasses, so should result in little additional friction for users to upgrade versions. Or we could make a (slightly) bigger jump to pydantic models

An example of deserialisation with pydantic models:

```
from pydantic import BaseModel

class A(BaseModel):
    foo: str
        
class B(BaseModel):
    a: A
    bar: str
    
b = B(**{'a': {'foo': '123'}, 'bar': 'abc'})
```

An example of deserialisation with pydantic dataclasses

```
from pydantic.dataclasses import dataclass

@dataclass
class A:
    foo: str
        
@dataclass
class B:
    a: A
    b: str

b = B(**{'a': {'foo': '123'}, 'bar': 'abc'})

```

**Comment by gr8jam** [comment]
Hello 👋 A fellow user of `ocpp` package here. I agree with the OP. Pydantic classes would be a really good fit for this project. 

Reimplementing its own serialization/deserialization logic instead of using a well established package like `pydantic` is error prone. For example, in version `1.0.0` a [bug](https://github.com/mobilityhouse/ocpp/issues/635) was introduced during the serialization of payload (here the [fix](https://github.com/mobilityhouse/ocpp/pull/637) the addressed this bug). 

If the maintainers of this project are intrigued by this idea and just lack the time for implementation, I would volunteer to make this change. 🙂
